### PR TITLE
decouple API from multiple backend on createMPU

### PR DESCRIPTION
**Pull request by ploki**
refactoring evaluation

### CODE REVIEW

File: lib/api/initiateMultipartUpload.js [hunk 1]
```diff
@@ -177,48 +172,15 @@ function initiateMultipartUpload(authInfo, request, log, callback) {
...
-                    uploadId = dataBackendResObj.UploadId;
-                } else {
-                    uploadId = UUID.v4().replace(/-/g, '');
-                }
+        const vcfg = destinationBucket.getVersioningConfiguration();
+        const isVersionedObj = vcfg && vcfg.Status === 'Enabled';
+        return data.createMPU(
+            objectKey, metaHeaders, bucketName, websiteRedirectHeader,
+            locConstraint, isVersionedObj, log,
+            (err, uploadId) => {
```
Comment (LaurenSpiegel) [inline lib/api/initiateMultipartUpload.js]: I would generate the upload id here if it wasn't returned from the backend.

File: lib/data/wrapper.js [hunk 2]
```diff
@@ -285,6 +288,69 @@ const data = {
         }
         return client.getDiskUsage(log.getSerializedUids(), cb);
     },
+
+    /**
+     * Initiate a multipart upload in the underlying storage
+     * @param {string} key - the key of the MPU object
+     * @param {object} metaHeaders - user defined metadata headers
+     * @param {string} bucketName - destination bucket name
+     * @param {XXX} websiteRedirectHeader - XXX
```
Comment (LaurenSpiegel) [inline lib/data/wrapper.js]: very explicit

File: lib/data/wrapper.js [hunk 3]
```diff
@@ -285,6 +288,69 @@ const data = {
...
+     * @param {object} log - request logger
+     * @param {function} callback - parameter list is (err, uploadId)
+     * @return {undefined}
+     */
+    createMPU:
+    (key, metaHeaders, bucketName, websiteRedirectHeader,
+     locConstraint, isVersionedObj, log, callback) => {
+        if (!client.createMPU) {
+            // if the backend doesn't implement this function, it is
+            // assumed the underlying storage would traduce this operation
```
Comment (LaurenSpiegel) [inline lib/data/wrapper.js]: translate rather than traduce

File: lib/data/wrapper.js [hunk 4]
```diff
@@ -285,6 +288,69 @@ const data = {
...
+                        .customizeDescription(externalVersioningErrorMessage));
+                    }
+                }
+                let uploadId;
+                if (dataBackendResObj) {
+                    // dataBackendResObj will be returned in data backend
+                    // handles mpu
+                    uploadId = dataBackendResObj.UploadId;
+                } else {
+                    uploadId = UUID.v4().replace(/-/g, '');
```
Comment (LaurenSpiegel) [inline lib/data/wrapper.js]: let's get this out of here.

File: lib/data/wrapper.js [hunk 5]
```diff
@@ -285,6 +288,69 @@ const data = {
...
+                        log.debug(externalVersioningErrorMessage,
+                            { method: 'multipleBackendGateway',
+                                error: errors.NotImplemented });
+                        return callback(errors.NotImplemented
+                        .customizeDescription(externalVersioningErrorMessage));
+                    }
+                }
+                let uploadId;
+                if (dataBackendResObj) {
+                    // dataBackendResObj will be returned in data backend
```
Comment (LaurenSpiegel) [inline lib/data/wrapper.js]: if there is a dataBackendResObj, it means the backend handled mpu

**Comment by ploki** [comment]
dear reviewer, only the last commit is relevant for this PR

File: lib/data/wrapper.js [hunk 6]
```diff
@@ -457,6 +457,65 @@ const data = {
...
+                  constants.externalBackends[config
+                      .locationConstraints[locConstraint].type]
+                ) {
+                    const externalVersioningErrorMessage =
+                              'We do not currently support putting ' +
+                              'a versioned object to a location-constraint ' +
+                              'of type AWS or Azure.';
+                    if (isVersionedObj) {
+                        log.debug(externalVersioningErrorMessage,
+                            { method: 'multipleBackendGateway',
```
Comment (dora-korpar) [inline lib/data/wrapper.js]: shouldn't this method be 'initiate mpu'?

**Comment by dora-korpar** [comment]
what's going to happen with the first commit?

**Pull request closed without merging**